### DISCUSSION TITLE
Static page caching and multiple forms on the same page

### DISCUSSION THREAD

**Discussion by icreatestuff** [solspace/craft-freeform#777] [Questions & Answers] [answered]
Hello,

We use Freeform and the Blitz plugin extensively across a couple of our sites.

We've followed the [templating guides here](https://docs.solspace.com/craft/freeform/v4/templates/caching/#static-page-caching-cdn-blitz) regarding how to work with cached forms to refresh the hash, honeypot and CSRF tokens which for the most part seems to be working great.

However, there are still some occasions where we do seem to run into the `400` error exception with `Unable to verify your data submission` message so I wanted to double-check what the process should be for refreshing the form data in those instances.

As far as we can tell the problem occurs when there are multiple forms on the same page and when the same form is on the page more than once. For example we have a donations form in the main header/navigation of the website as well as in the footer.

We have the JS enhancement setting enabled. Each form on the page gets a `js-freeform` class.
We have followed the instructions  to set `id` and `fieldIdPrefix` [as shown here ](https://docs.solspace.com/craft/freeform/v4/guides/multiple-instances-of-same-form/#multiple-instances-of-the-same-form).

Our javascript for refreshing the data on page load looks like this:

```js
const forms = document.querySelectorAll(".js-freeform");

if (forms.length) {
    forms.forEach((form) => {
        // Refresh CSRF and Hash
        axios
            .get(`/ajax/freeform?form=${form.dataset.handle}`, {
                headers: { "X-Requested-With": "XMLHttpRequest" },
            })
            .then((response) => {
                // Update the Form Hash
                form.querySelector("input[name=formHash]").value = response.data.hash;

                // Update the Payload if encrypted payloads are enabled
                form.querySelector("input[name=freeform_payload]").value = response.data.payload;

                // Update the JS Honeypot (ONLY when JS Enhancement setting is ON)
                var honeypot = response.data.honeypot;
                var honeypotInput = form.querySelector('input[name^="freeform_form_handle"]');

                honeypotInput.setAttribute("id", honeypot.name);
                honeypotInput.setAttribute("name", honeypot.name);
                honeypotInput.value = honeypot.hash;

                // Locate and update the CSRF input
                var csrf = response.data.csrf;
                form.querySelector(`input[name=${csrf.name}]`).value =
                    csrf.value;
            });
    });
}
```

The `/ajax/freeform` template is excluded from caching and looks like this:

```twig
{% set form = craft.freeform.form(craft.app.request.get('form')) %}
{{ {
    hash: form.hash,
    honeypot: craft.freeform.honeypot(form),
    payload: form.payload,
    csrf: {
        name: craft.app.config.general.csrfTokenName,
        value: craft.app.request.csrfToken,
    }
}|json_encode|raw }}
```

The javascript works and the template returns what we would expect. Is there anything we could be missing or have setup correctly? For example, should the identical forms don't he page be given the same hash, honeypot and CSRF values or is it ok that they are different per form?

All help is appreciated, many thanks.

**Comment by kjmartens** [comment]
Hi @icreatestuff,

Can you first confirm which exact version of Freeform and Craft you're using? 🙂

**Comment by icreatestuff** [comment]
Hi @kjmartens yes of course, should have thought about that!

Craft Pro 4.4.17
Freeform 4.1.6

**Comment by gustavs-gutmanis** [comment, marked as answer]
Hi @icreatestuff,

I have a hunch about a possible issue.
Since you create a new request per form, there is a possibility that one of the first CSRF tokens would get invalidated, when requesting several at a time.

Could you leave everything as is, but for the CSRF part - always replace every form's CSRF token with the latest one you receive?
The CSRF token can be the same for all forms, but the rest of the hashes and tokens should be different.

Hope this helps.

**Reply by icreatestuff** [reply]
Thanks for the suggestion @gustavs-gutmanis I'll try that out.

Out of curiosity would this approach allow someone to submit form 1 AND form 2 on the same page, assuming there's no page refresh or updating of the CSRF token after submitting form 1?

**Reply by gustavs-gutmanis** [reply]
In theory it should, yes.
